docs/source/examples/replicate_vicuna/chatbot.py

```diff
@@ -0,0 +1,110 @@
+import ctranslate2
+import gradio as gr
+import os
+import sentencepiece as spm
+import time
+
+CACHE = {}
+model_dir = "finetuned_llama7B/llama7B-vicuna-onmt_step_4000.concat_CT2"
+tokenizer_dir = "llama"
+out_dir = "outputs/ct2"
+
+
+def load_models(model_dir, tokenizer_dir):
+    if CACHE.get("generator", None) is None:
+        CACHE["generator"] = ctranslate2.Generator(model_dir, device="cuda")
+        CACHE["tokenizer"] = spm.SentencePieceProcessor(
+            os.path.join(tokenizer_dir, "tokenizer.model")
+        )
+
+
+def make_prompt(chat_history):
+    task_description = "Below is an instruction that describes a task. Write a response that appropriately completes the request.｟newline｠｟newline｠"  # noqa:E501
+
+    def parse_instruction(text):
+        return f"### Instruction:｟newline｠ {text} ｟newline｠｟newline｠"
+
+    def parse_response(text):
+        return f"### Response:｟newline｠{text}"
+
+    out = [task_description]
+    for _user_message, _bot_message in chat_history:
+        out.append(parse_instruction(_user_message))
+        if _bot_message is not None:
+            out.append(parse_response(_bot_message))
+        else:
+            out.append("### Response:｟newline｠")
+    prompt = "".join(out)
+    return prompt
+
+
+def generate_words(generator, sp, prompt, add_bos=True):
+    prompt_tokens = sp.encode(prompt, out_type=str)
+
+    if add_bos:
+        prompt_tokens.insert(0, "<s>")
+
+    step_results = generator.generate_tokens(
+        prompt_tokens, sampling_temperature=0.1, sampling_topk=40, max_length=512
+    )
+
+    output_ids = []
+
+    for step_result in step_results:
+        is_new_word = step_result.token.startswith("▁")
+
+        if is_new_word and output_ids:
+            yield " " + sp.decode(output_ids)
+            output_ids = []
+
+        output_ids.append(step_result.token_id)
+
```

Comment:
with CT2 it does not detokenize the ((newline)) token so you need to do it somewhere.